crates/codegen/schema/src/validation/ast/productions.rs

```diff
@@ -0,0 +1,214 @@
+use indexmap::IndexMap;
+use semver::Version;
+
+use crate::{ast_optional, ast_value, types, validation::ast::Node, yaml::cst};
+
+pub type ProductionRef = std::rc::Rc<Production>;
+
+pub struct Production {
```

Comment:
most of the AST conversions here are strong-typed (love rust macros!), except the two enums below, because they are waiting on https://github.com/rust-lang/rust/issues/86935

Will implement as well once that issue is stabilized.